bin/brew

```diff
@@ -45,5 +85,77 @@ fi
 
 export HOMEBREW_BREW_FILE
 export HOMEBREW_RUBY_PATH
+export HOMEBREW_PREFIX
+export HOMEBREW_REPOSITORY
+export HOMEBREW_LIBRARY
+export HOMEBREW_CELLAR
+
+if [ -n "$HOMEBREW_OSX" ]
+then
+  if [ "$(xcode-\select --print-path)" = "/" ]
+  then
+    cat >&2 <<EOS
+Your xcode-select path is currently set to '/'.
+This causes the 'xcrun' tool to hang, and can render Homebrew unusable.
+If you are using Xcode, you should:
+  sudo xcode-select -switch /Applications/Xcode.app
+Otherwise, you should:
+  sudo rm -rf /usr/share/xcode-select
+EOS
+    exit 1
+  fi
+
+  XCRUN_OUTPUT="$(/usr/bin/xcrun clang 2>&1)"
+  XCRUN_STATUS="$?"
+
+  if [ "$XCRUN_STATUS" != "0" ] && [[ "$XCRUN_OUTPUT" =~ license ]]
+  then
+    cat >&2 <<EOS
+You have not agreed to the Xcode license. Please resolve this by running:
+  sudo xcodebuild -license
+EOS
+    exit 1
+  fi
+fi
+
+# Many Pathname operations use getwd when they shouldn't, and then throw
+# odd exceptions. Reduce our support burden by showing a user-friendly error.
+if ! [ -d "$(pwd)" ]
+then
+  echo "The current working directory doesn't exist, cannot proceed."
+  exit 1
+fi
+
+for i in "$@"
+do
+  [[ "$i" =~ ^- ]] && continue
+  HOMEBREW_COMMAND="$i"
+  HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$i.sh"
+  break
+done
+
+if [ "$(id -u)" = "0" ] && [ "$(stat -f%u "$HOMEBREW_BREW_FILE")" != "0" ]
+then
+  for cmd in install reinstall postinstall link pin unpin update update-bash upgrade create migrate tap switch
+  do
+    [ "$HOMEBREW_COMMAND" != "$cmd" ] && continue
+    cat >&2 <<EOS
+Cowardly refusing to 'sudo brew $i'
+You can use brew with sudo, but only if the brew executable is owned by root.
+However, this is both not recommended and completely unsupported so do so at
+your own risk.
+EOS
+    exit 1
+  done
+fi
 
-exec "$HOMEBREW_RUBY_PATH" -W0 "$BREW_LIBRARY_DIRECTORY/brew.rb" "$@"
+if [ -n "$HOMEBREW_BASH_COMMAND" ] && [ -x "$HOMEBREW_BASH_COMMAND" ]
+then
+  # eval rather than executing directly to ensure the entire file is read into
+  # memory before it is run. This makes running a Bash script behave more like
+  # a Ruby script and avoids hard-to-debug issues if the Bash script is updated
+  # at the same time as being run.
+  eval "$(cat "$HOMEBREW_BASH_COMMAND")" "$@"
```

Comment:
This works for `update-bash` because that one either exits via `exit` or via `exec` and never returns. To better support future shell-only commands that could simply return, I would suggest to move the remaining top-level code in the command into a function named `brew_run_command` which would look like this for `update-bash.sh`:

```shell
brew_run_command() {
  update "$@"
  exec "$HOMEBREW_BREW_FILE" update-bash-report "$@"
}
```

and then replace this line with the following two lines:

```shell
source "$HOMEBREW_BASH_COMMAND"
{ brew_run_command "$@" ; exit $? ; }
```

This will allow a command implemented in shell to return via `return <exit_code>` and also more closely matches the way internal commands are implemented in Ruby. (The grouping still makes sure that nothing else is executed, even if `bin/brew` itself is modified during an update.)

----

If you really want to stick with `eval` (I hope not), then `"$(cat "$HOMEBREW_BASH_COMMAND")"` can be simplified to `"$(<"$HOMEBREW_BASH_COMMAND")"`, i.e. no `cat` is needed.